src/mdio/schemas/v1/templates/abstract_dataset_template.py

```diff
@@ -0,0 +1,152 @@
+        # e.g. ["cdp-x", "cdp-y"] for 3D pre-stack CPD depth
+        # e.g. ["gun", "shot-x", "shot-y", "receiver-x", "receiver-y"] for 3D pre-stack
+        #      time Shot gathers
+        self._coord_names = []
+        # Name of the variable in the dataset
+        # e.g. "StackedAmplitude" for 2D post-stack depth
+        # e.g. "StackedAmplitude" for 3D post-stack depth
+        # e.g. "AmplitudeCDP" for 3D pre-stack CPD depth
+        # e.g. "AmplitudeShot" for 3D pre-stack time Shot gathers
+        self._var_name = ""
+        # Chunk shape for the variable in the dataset
+        # e.g. [1024, 1024] for 2D post-stack depth
+        # e.g. [128, 128, 128] for 3D post-stack depth
+        # e.g. [1, 1, 512, 4096] for 3D pre-stack CPD depth
+        # e.g. [1, 1, 512, 4096] for 3D pre-stack time Shot gathers
+        self._var_chunk_shape = []
+
+        # Variables instantiated when build_dataset() is called
+        self._builder: MDIODatasetBuilder = None
+        # Sizes of the dimensions in the dataset, to be set when build_dataset() is called
+        self._dim_sizes = []
+        # Horizontal units for the coordinates (e.g, "m", "ft"), to be set when
+        # build_dataset() is called
+        self._coord_units = []
+
+    def build_dataset(self, name: str, sizes: list[int], coord_units: list[AllUnits]) -> Dataset:
+        """Template method that builds the dataset.
+
+        Args:
+            name: The name of the dataset.
+            sizes: The sizes of the dimensions.
+            coord_units: The units for the coordinates.
+
+        Returns:
+            Dataset: The constructed dataset
+        """
+        self._dim_sizes = sizes
+        self._coord_units = coord_units
+
+        self._builder = MDIODatasetBuilder(name=name, attributes=self._load_dataset_attributes())
+        self._add_dimensions()
+        self._add_coordinates()
+        self._add_variables()
+        return self._builder.build()
+
+    def get_name(self) -> str:
+        """Returns the name of the template."""
+        return self._get_name()
+
+    @abstractmethod
+    def _get_name(self) -> str:
+        """Abstract method to get the name of the template.
+
+        Must be implemented by subclasses.
+
+        Returns:
+            str: The name of the template
+        """
+
+    @abstractmethod
+    def _load_dataset_attributes(self) -> UserAttributes:
+        """Abstract method to load dataset attributes.
+
+        Must be implemented by subclasses.
+
+        Returns:
+            UserAttributes: The dataset attributes
+        """
+
+    def _add_dimensions(self) -> None:
+        """A virtual method that can be overwritten by subclasses to add custom dimensions.
+
+        Uses the class field 'builder' to add dimensions to the dataset.
+        """
+        for i in range(len(self._dim_names)):
+            self._builder.add_dimension(self._dim_names[i], self._dim_sizes[i])
+
+    def _add_coordinates(self) -> None:
```

Comment:
It is not immediately obvious what the default behavior of this method is. Should we make it into an abstract method and enforce people to add the coordinates? It would also clean up the `AbstractDatasetTemplate.__init__`. Thoughts?

Reply:
i think its ok with float64 for now. hopefully these won't be overridden that much!